_posts/2020-08-28-porting-namecoin-in-tor-browser-to-windows.md

```diff
@@ -0,0 +1,108 @@
+(The latter took some DuckDuckGo-fu, as it turns out that `sys.exit()` can't be called from a child thread in Python; the correct way to exit from a child thread is `os._exit()`.)
+
+Some testing revealed that this worked; I was able to remove the Bash code that terminated Electrum-NMC and StemNS, and they still exited properly.  
+Moving on...
+
+## Launching Namecoin when Tor Browser Starts
+
+Of course, the other part of the Bash launcher for Namecoin was the code that launches Electrum-NMC and StemNS.  
+For this, I ended up copying/pasting the code in Tor Launcher that launches the Tor daemon.  
+As XPCOM-based JavaScript code goes, Tor Launcher is pretty readable, so the copy/paste job wasn't particularly eventful.  
+I *did* notice that the documentation for debugging Tor Launcher was outdated, but the Tor developers on IRC were able to point me in the right direction there.
+
+I won't bore you with too many details on this part; it was mostly grunt work.  
+But I got it working.  
+Excellent, moving on....
+
+## Windows and Python
+
+By this point, all of the GNU/Linux-specific code had been replaced with cross-platform code.  
+So we were ready to move onto enabling Windows support.  
+Most of this was as simple as tweaking the rbm descriptors to enable Namecoin on Windows, and fixing a few bugs where the Go dependencies for certain rbm projects were wrong on Windows.  
+But, there was one issue that needed dealing with: Python.
+
+Both Electrum-NMC and StemNS are written in Python.  
+GNU/Linux systems generally have Python available by default, but this is not the case for Windows.  
+In addition, on GNU/Linux, Python scripts are executable programs, but on Windows, they're data files that need to be explicitly opened with the Python interpreter.  
+This means that some Windows-specific things need to be done.
+
+For Python being available, I took the easy way out: I simply defined it to be out of scope, i.e. the user is responsible for installing Python themselves before enabling Namecoin in Tor Browser.  
+I also rigged Tor Launcher to check every directory in the `PATH` for a Python interpreter, and to use the discovered Python binary as the executable, passing the Electrum-NMC or StemNS path as a command-line argument instead.  
+Worked pretty well.
+
+Finally, I re-implemented verbose logging for Namecoin (which was also part of the Bash code that I had removed).  
+This was done via a `TOR_VERBOSE_NAMECOIN` environment variable.  
+It does two things when enabled:
+
+1. Pass the `-v` argument to Electrum-NMC, which enables verbose output.
+2. On Windows, use `python.exe` (which pops up a command prompt window with logs) instead of `pythonw.exe` (which doesn't spawn a command window).
+
+Oh, and I also had to tweak the linker flags for `ncprop279` to make it avoid launching a command prompt window on Windows as well.
```

Comment:
"as well" duplicates "also" for a strange flow.
(Nit) "Oh" is unclear as to whether it is aimed at making the sentence more or less important.
> I also had to tweak the linker flags for `ncprop279` to make it avoid launching a command prompt window on Windows.
> I had to tweak the linker flags for `ncprop279` to make it avoid launching a command prompt window on Windows as well.